packages/daemon/DESIGN.md

```diff
@@ -0,0 +1,215 @@
+
+Endo is a user agent for running guest programs.
+These programs receive capabilities from the user in the form of references to
+potentially powerful local or remote objects.
+Endo uses a system called Eventual Send that that stretches references to
+remote objects over one or more capability transfer protocols (CapTP).
+Endo is an example of the Powerbox Pattern[^powerbox].
+
+[^powerbox]: See _The Powerbox Pattern_ from [How Emily Tamed the
+    Caml][powerbox].
+
+[powerbox]: https://www.hpl.hp.com/techreports/2006/HPL-2006-116.html
+
+A Promise represents a result from either the past or the future.
+Eventual Send displaces a result in both space and time.
+That is, a program can use eventual send to uniformly interact with a result
+regardless of whether the eventual result is local or remote, pending or
+settled.
+Furthermore, Eventual Send can pipeline[^pipeline] messages to an eventual result,
+like invoking a method on the result of a prior method invocation, to avoid
+unnecessary round trips through the underlying CapTP connection.
+
+[^pipeline]: See [Mark's explanation of pipelines][pipeline-erights], the
+    [Cap'n Proto explanation][pipeline-capnproto], [Wikipedia][pipeline-wikipedia], or
+    the [ECMAScript standard proposal][pipeline-proposal].
+
+[pipeline-erights]: http://erights.org/elib/distrib/pipeline.html
+[pipeline-capnproto]: https://capnproto.org/rpc.html
+[pipeline-wikipedia]: https://en.wikipedia.org/wiki/Futures_and_promises#Promise_pipelining
+[pipeline-proposal]: https://www.proposals.es/proposals/Support%20for%20Distributed%20Promise%20Pipelining
+
+Much hinges on what happens when a program restarts.
+Like most processes, Endo workers start without any memory from their previous
+run.
+So, Endo programs must manually persist[^persistence] anything they wish to
+remember after a restart.
+
+[^persistence]: Aside: Manual persistence is as opposed to Orthogonal
+    Persistence, where a program resumes exactly where it left off when it
+    restarts, albeit it by replaying every message it recieved since the last
+    snapshot or inception.  The architecture of Endo allows for Orthogonal
+    Persistence to be built on top of a manually persisted worker and a durable
+    reference to external storage for snapshots, message log, and virtual
+    objects.
+
+Endo applications ultimately receive all their data and powers from the Endo
+daemon with the user's consent.
+Endo persists **data** between restarts with a content address store in the
+user's home directory.
+Endo persists **powers** between restarts by storing formulas in the user's
+home directory.
+A formula describes how Endo should recreate a value, including powerful
+capability bearing objects if it restarts.
+
+When Endo runs a confined program, it provides a Powers object that the
```

Comment:
> When Endo runs a confined program

Does "Endo application" as used here === "A confined program that Endo runs" === "A confined program that the Endo daemon runs"?

Reply:
The nesting looks like:

E. Daemon > E. Worker > E. Compartment > E. Application

As above, I would like to entertain Extension or Caplet instead of Application.

Reply:
What does "X > Y" mean here? I tried on
   * cardinality: there are multiple Xs per Y
   * taxonomy: every X is also a Y

and neither seemed to fit.

Reply:
Generally, “contains”

- Endo Daemon spawns Endo workers
- Endo Workers construct Endo Compartments
- Endo Compartments confine Endo Applications (extensions, caplets, guest programs)